halo2_gadgets/src/ecc.rs

```diff
@@ -259,7 +268,7 @@ impl<C: CurveAffine, EccChip: EccInstructions<C>> NonIdentityPoint<C, EccChip> {
     pub fn mul(
         &self,
         mut layouter: impl Layouter<C::Base>,
-        by: &EccChip::Var,
+        by: &EccChip::ScalarVar,
     ) -> Result<(Point<C, EccChip>, ScalarVar<C, EccChip>), Error> {
```

Comment:
```suggestion
        by: &mut ScalarVar<C, EccChip>,
    ) -> Result<Point<C, EccChip>, Error> {
```

or, if we take @ebfull's preferred [move-based API approach](https://github.com/zcash/halo2/issues/483#issuecomment-1072587613):
```suggestion
        by: ScalarVar<C, EccChip>,
    ) -> Result<(Point<C, EccChip>, ScalarVar<C, EccChip>), Error> {
```

Both of these require adding new constructors to `ScalarVar`:
```rust
impl<C: CurveAffine, EccChip: EccInstructions<C>> ScalarVar<C, EccChip> {
    /// Witnesses the given full-width scalar.
    ///
    /// Depending on the `EccChip` implementation, this may either witness the scalar
    /// immediately, or delay witnessing until its first use in [`NonIdentityPoint::mul`].
    pub fn new(
        chip: EccChip,
        mut layouter: impl Layouter<C::Base>,
        value: Option<C::Scalar>,
    ) -> Self {
        let scalar = chip.witness_scalar_var(&mut layouter, value);
        scalar.map(|inner| ScalarVar { chip, inner })
    }
}

impl<C: CurveAffine, EccChip: BaseFitsInScalarInstructions<C>> ScalarVar<C, EccChip> {
    /// Constructs a scalar from an existing base-field element.
    pub fn from_base(
        chip: EccChip,
        mut layouter: impl Layouter<C::Base>,
        base: &Self::Var,
    ) -> Self {
        let scalar = chip.scalar_var_from_base(&mut layouter, value);
        scalar.map(|inner| ScalarVar { chip, inner })
    }
}
```

We also need to check that the scalar and point chips match.